pkg/infra/x/nanogit/listing.go

```diff
@@ -0,0 +1,284 @@
+	trees := make(map[string]*node, 0)
+	entries := make(map[string]*node, 0)
+	for i, p := range parts {
+		if string(p) == "packfile\n" {
+			file := readPackfileResponse(parts, i)
+
+			rr, err := NewPackfileReader(bytes.NewReader(file), 1000000, nil)
+			if err != nil {
+				return nil, err
+			}
+			p, err := rr.Read()
+			if err != nil {
+				return nil, err
+			}
+
+			for name, t := range p.Trees {
+				n := &node{
+					Hash:     name,
+					IsFolder: true,
+				}
+				trees[n.Hash] = n
+				for _, e := range t.Entries {
+					child := &node{
+						Name: e.Name,
+						Hash: e.Hash,
+					}
+					n.Children = append(n.Children, child)
+
+					old := entries[e.Hash]
+					if old != nil {
+						old.IsShared = true
+						child.IsShared = true
+						shared = append(shared, old, child)
+					}
+					entries[e.Hash] = child // This may be a tree!
+				}
+			}
+		}
+	}
+
+	for _, tree := range trees {
+		entry, ok := entries[tree.Hash]
+		if ok {
+			if entry.Children != nil {
+				listing.Errors = append(listing.Errors, fmt.Sprintf("found twice %s", tree.Hash))
+			}
+			entry.IsFolder = true
+			if entry.IsShared {
+				tree.sort()
+				if listing.Shared == nil {
+					listing.Shared = make(map[string][]*node)
+				}
+				listing.Shared[tree.Hash] = tree.Children
+			} else {
+				entry.Children = tree.Children
+			}
+
+		} else if listing.Root == nil {
+			listing.Root = tree
+		} else {
+			listing.Errors = append(listing.Errors, fmt.Sprintf("found multiple roots: %s", tree.Hash))
+			fmt.Printf("multiple roots???")
+		}
+
+		if tree.IsShared {
+			fmt.Printf("SHARED TREEXX: %s\n", tree.Hash)
+		}
+
+	}
+
+	if listing.Shared != nil {
+		for _, entry := range shared {
+			if listing.Shared[entry.Hash] != nil {
+				entry.IsFolder = true
+				entry.IsShared = true
+			}
+		}
+	}
+
+	listing.Root.sort()
+	frame := listing.toDataFrame()
+	frame.Meta = &data.FrameMeta{
+		Custom: &GetFrameMeta{
+			Address:   addr,
+			BytesRead: len(x),
+		},
+	}
+	return frame, err
+}
+
+// https://git-scm.com/docs/protocol-v2
+func ReadBody(addr GitAddress, oids ...string) error {
+	lines := []string{
+		"command=fetch\n",
+		"object-format=sha1\n",
+		"",
+		"thin-pack\n",
+		"no-progress\n",
+		"ofs-delta\n",
+		"deepen 1\n",
+		"filter blob:none\n",
+		"shallow " + addr.Branch + "\n",
+	}
+	for _, s := range oids {
+		lines = append(lines, fmt.Sprintf("want %s\n", s))
+	}
+	lines = append(lines, "done\n")
+
+	z, err := cmd(addr.Owner, addr.Repo, fmtLines(lines))
+	if err != nil {
+		return err
+	}
+	parts, _ := parsePktLine(z)
+	for i, p := range parts {
+		if string(p) == "packfile\n" {
+			file := readPackfileResponse(parts, i)
+
+			cache := map[string][]byte{}
+			rr, _ := NewPackfileReader(bytes.NewReader(file), 1000000, func(hash string, content []byte) {
+				cache[hash] = content
+			})
+			p, err := rr.Read()
+			if err != nil {
+				return err
+			}
+
+			for k, v := range p.Commits {
+				fmt.Println("Commit", k, v.Author, v.Message)
```

Comment:
@zserge -- this never has commits, any idea how to add commit info in the response? or does it have to come from another request